examples/python/cancellation/search.py

```diff
@@ -0,0 +1,158 @@
+# Copyright the 2019 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A search algorithm over the space of all bytestrings."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import base64
+import hashlib
+import logging
+import struct
+
+from examples.python.cancellation import hash_name_pb2
+
+_LOGGER = logging.getLogger(__name__)
+_BYTE_MAX = 255
+
+
+def _get_hamming_distance(a, b):
+    """Calculates hamming distance between strings of equal length."""
+    distance = 0
+    for char_a, char_b in zip(a, b):
+        if char_a.lower() != char_b.lower():
+            distance += 1
+    return distance
+
+
+def _get_substring_hamming_distance(candidate, target):
+    """Calculates the minimum hamming distance between between the target
+        and any substring of the candidate.
+
+    Args:
+      candidate: The string whose substrings will be tested.
+      target: The target string.
+
+    Returns:
+      The minimum Hamming distance between candidate and target.
+    """
+    min_distance = None
+    for i in range(len(candidate) - len(target) + 1):
+        distance = _get_hamming_distance(candidate[i:i + len(target)], target)
+        if min_distance is None or distance < min_distance:
+            min_distance = distance
+    return min_distance
+
+
+def _get_hash(secret):
+    hasher = hashlib.sha1()
+    hasher.update(secret)
+    return base64.b64encode(hasher.digest()).decode('ascii')
+
+
+class ResourceLimitExceededError(Exception):
+    """Signifies the request has exceeded configured limits."""
+
+
+def _bytestrings_of_length(length):
+    """Generates a stream containing all bytestrings of a given length.
+
+    Args:
+      length: A positive integer length.
+
+    Yields:
+      All bytestrings of length `length`.
+    """
+    digits = [0] * length
```

Comment:
There is a fancy way to do this through `itertools`:

```Python
def _bytestrings_of_length(length):
    product = itertools.product(range(256), repeat=length)
    while True:
        yield b''.join(struct.pack('B', i) for i in next(product))
```

Reply:
Nice!